src/io/flutter/run/test/FlutterTestLineMarkerContributor.java

```diff
@@ -20,25 +29,94 @@
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+import javax.swing.*;
+import java.util.Date;
+import java.util.Map;
+
 
 public class FlutterTestLineMarkerContributor extends RunLineMarkerContributor {
+
+  private static final int TEST_LIMIT = 1024;
+
   @Nullable
   @Override
   public Info getInfo(@NotNull PsiElement element) {
     if (isTestCall(element)) {
+      final Icon icon = getTestStateIcon(element);
       final AnAction[] actions = ExecutorAction.getActions();
       final Function<PsiElement, String> tooltipProvider =
         psiElement -> StringUtil.join(ContainerUtil.mapNotNull(actions, action -> getText(action, element)), "\n");
-      return new Info(AllIcons.RunConfigurations.TestState.Run, tooltipProvider, actions);
+      return new Info(icon, tooltipProvider, actions);
     }
 
     return null;
   }
 
+  private static Icon getTestStateIcon(@NotNull PsiElement element) {
+
+    // SMTTestProxy maps test run data to a URI derived from a location hint produced by `package:test`.
+    // If we can find corresponding data, we can provide state-aware icons.  If not, we default to
+    // a standard Run state.
+
+    PsiFile containingFile;
+    try {
+      containingFile = element.getContainingFile();
+    }
+    catch (PsiInvalidElementAccessException e) {
+      containingFile = null;
+    }
+
+    final Project project = element.getProject();
+    final PsiDocumentManager psiDocumentManager = PsiDocumentManager.getInstance(project);
+
+    final Document document = containingFile == null ? null : psiDocumentManager.getDocument(containingFile);
+    if (document != null) {
+      final int textOffset = element.getTextOffset();
+      final int lineNumber = document.getLineNumber(textOffset);
+
+      // e.g., dart_location:///Users/pq/IdeaProjects/untitled1298891289891/test/unit_test.dart,3,2,["my first unit test"]
+      final String path = containingFile.getVirtualFile().getPath();
+      final String testLocationPrefix = "dart_location://" + path + "," + lineNumber;
+
+      final TestStateStorage storage = TestStateStorage.getInstance(project);
+      if (storage != null) {
+        final Map<String, TestStateStorage.Record> tests =
+          storage.getRecentTests(TEST_LIMIT, getSinceDate());
+        if (tests != null) {
+          for (Map.Entry<String, TestStateStorage.Record> entry : tests.entrySet()) {
```

Comment:
It looks like this is iterating over all recent tests (up to TEST_LIMIT and within the last day) once per PsiElement that matches isTestCall(). It seems like it might be expensive in a project has has many test files open?